### PR TITLE
diff: complete support for git patchid

### PR DESCRIPTION
**Pull request by herrerog**
Hi,
I noticed that git_diff_patchid wasn't returning a correct hash in some cases.
So I fixed those and added some tests for it.
Please let me know what you think.
Thanks!

### CODE REVIEW

**Comment by ethomson** [comment]
Overall, this looks very nice to me.  Thanks for the PR!

/cc @pks-t to look at some of the factoring, especially since patchid was his work originally and he has more familiarity.

Note that the build failures appear to be due to #5308.

File: src/diff.c [hunk 1]
```diff
@@ -426,81 +426,39 @@ static void strip_spaces(git_buf *buf)
...
-		return -1;
-	}
-
-	git_buf_put(&buf, line->content, line->content_len);
 	strip_spaces(&buf);
 
-	if ((error = git_hash_update(&args->ctx, buf.ptr, buf.size)) < 0)
+	if (line->origin == GIT_DIFF_LINE_FILE_HDR && !args->first_file)
+		if ((error = (flush_hunk(&args->result, &args->ctx))) < 0)
+			goto out;
```
Comment (pks-t) [inline src/diff.c]: Stylistic nit: the nested if isn't required here and we usually use `&&` chaining in these cases. It's a matter of taste though, so feel free to ignore this comment.
Please remove the braces around `flush_hunk`, though.

File: src/diff.c [hunk 2]
```diff
@@ -526,7 +484,10 @@ int git_diff_patchid(git_oid *out, git_diff *diff, git_diff_patchid_options *opt
 	if ((error = git_hash_ctx_init(&args.ctx)) < 0)
 		goto out;
 
-	if ((error = git_diff_foreach(diff, file_cb, NULL, NULL, patchid_line_cb, &args)) < 0)
+	if ((error = (git_diff_print(diff,
+				     GIT_DIFF_FORMAT_PATCH_ID,
+				     git_diff_patchid_print_callback__to_buf,
+				     &args))) < 0)
```
Comment (pks-t) [inline src/diff.c]: I like your approach, it's a lot simpler than what I was doing before.

File: src/diff.c [hunk 3]
```diff
@@ -426,81 +426,39 @@ static void strip_spaces(git_buf *buf)
...
-
-	git_buf_put(&buf, line->content, line->content_len);
 	strip_spaces(&buf);
 
-	if ((error = git_hash_update(&args->ctx, buf.ptr, buf.size)) < 0)
+	if (line->origin == GIT_DIFF_LINE_FILE_HDR && !args->first_file)
+		if ((error = (flush_hunk(&args->result, &args->ctx))) < 0)
+			goto out;
+
+	if ((error = (git_hash_update(&args->ctx, buf.ptr, buf.size))) < 0)
```
Comment (pks-t) [inline src/diff.c]: Well, at least you're consistent :P Please follow libgit2's style and consistently remove braces around function invocations.

File: src/diff_print.c [hunk 4]
```diff
@@ -571,7 +571,7 @@ static int diff_print_patch_file(
 		(pi->flags & GIT_DIFF_FORCE_BINARY);
 	bool show_binary = !!(pi->flags & GIT_DIFF_SHOW_BINARY);
 	int id_strlen = pi->id_strlen;
-	bool print_index = true;
+	bool print_index = (pi->format != GIT_DIFF_FORMAT_PATCH_ID);
```
[the remark this answers is not in the document]
Comment (herrerog) [reply src/diff_print.c]: Thanks!

File: src/diff.c [hunk 5]
```diff
@@ -426,81 +426,39 @@ static void strip_spaces(git_buf *buf)
...
+	const git_diff_line *line,
 	void *payload)
 {
 	struct patch_id_args *args = (struct patch_id_args *) payload;
 	git_buf buf = GIT_BUF_INIT;
-	int error;
+	int error = 0;
 
-	GIT_UNUSED(progress);
+	git_buf_sanitize(&buf);
```
[the remark this answers is not in the document]
Comment (herrerog) [reply src/diff.c]: Right, I fixed that.

File: src/diff.c (hunk 1 above)
Comment (herrerog) [reply src/diff.c]: Ah good catch for the `&&`. This looks better.
I will remove all braces around functions calls.

File: src/diff.c (hunk 3 above)
Comment (herrerog) [reply src/diff.c]: Lol yeah, not sure why I was thinking this was the way to go for libgit2.

File: src/diff.c (hunk 2 above)
Comment (herrerog) [reply src/diff.c]: Thanks

**Comment by herrerog** [comment]
Thanks for your nice feedback, happy to help :)
I updated the branch.

**Comment by pks-t** [comment]
Thanks a lot for this well-crafted pull request, @herrerog!

**Pull request merged**